clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -204,9 +206,19 @@ public double compressionRate() {
     public boolean isFull() {
         return recordsBuilder.isFull();
     }
-
-    public void close() {
-        recordsBuilder.close();
+    /**
+     * Writes the final metadata in the batch and updates transaction state with the new sequence number for the
+     * TopicPartition
+     * @param transactionState the global transaction state which keeps track of sequence numbers for each TopicPartition.
+     */
+    synchronized public void close(TransactionState transactionState) {
+        if (!isClosed) {
+            recordsBuilder.close();
+            if (transactionState != null) {
+                transactionState.incrementSequenceNumber(topicPartition, this.recordCount);
```

Comment:
I originally started with having the callers of `close` do this. The major problem however, is that `close` can be called multiple times on the same batch. But we need to ensure that this executes exactly once. By putting this here, guarding by the `isClose` member variable, and making the method synchronized, we can get the exactly once processing in a clear way.